### PR TITLE
Get study by UUID 

### PR DESCRIPTION
**Pull request by ChamseddineBhd**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 1]
```diff
@@ -79,19 +78,19 @@ public StudyController(StudyService studyService) {
                                                   @RequestHeader("userId") String userId) {
         Mono<BasicStudyInfos> createStudy = studyService.createStudy(studyName, Mono.just(caseFile), description, userId, isPrivate)
                 .log(StudyService.ROOT_CATEGORY_REACTOR, Level.FINE);
-        return ResponseEntity.ok().body(studyService.assertStudyNotExists(studyName, userId).then(createStudy));
+        return ResponseEntity.ok().body(createStudy);
     }
 
-    @GetMapping(value = "/{userId}/studies/{studyName}")
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyController.java:85]: why still userId in the path ?

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 2]
```diff
@@ -102,308 +101,308 @@ public StudyController(StudyService studyService) {
         return ResponseEntity.ok().contentType(MediaType.APPLICATION_JSON).body(studyService.studyExists(studyName, userId));
     }
 
-    @DeleteMapping(value = "/{userId}/studies/{studyName}")
+    @DeleteMapping(value = "/{userId}/studies/{studyUuid}")
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyController.java:104]: same

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 3]
```diff
@@ -300,52 +308,55 @@ public StudyEntity doGetStudyWithPreFetchedLoadFlowResult(String studyName, Stri
...
+                                          LoadFlowResultEntity loadFlowResult, LoadFlowParametersEntity loadFlowParameters, UUID securityAnalysisUuid) {
+        return insertStudyEntity(uuid, studyName, userId, isPrivate, networkUuid, networkId, description, caseFormat, caseUuid, casePrivate, loadFlowStatus, loadFlowResult,
+                loadFlowParameters, securityAnalysisUuid)
+                .doOnSuccess(s -> emitStudyChanged(uuid, StudyService.UPDATE_TYPE_STUDIES));
     }
 
     private Mono<StudyCreationRequestEntity> insertStudyCreationRequest(String studyName, String userId, boolean isPrivate) {
         return insertStudyCreationRequestEntity(studyName, userId, isPrivate)
-                .doOnSuccess(s -> emitStudyChanged(studyName, StudyService.UPDATE_TYPE_STUDIES));
+                .map(studyCreationRequestEntity -> {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:356]: why change this ?

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 4]
```diff
@@ -617,48 +628,48 @@ public void doDeleteStudyIfNotCreationInProgress(String studyName, String userId
...
                 return Mono.empty();
             })
                     .then(monoUpdateLfState);
         });
     }
 
-    Mono<Void> runLoadFlow(String studyName, String userId) {
-        return setLoadFlowRunning(studyName, userId).then(getNetworkUuid(studyName, userId)).flatMap(uuid -> {
+    Mono<Void> runLoadFlow(UUID studyUuid) {
+        return setLoadFlowRunning(studyUuid).then(getNetworkUuid(studyUuid).flatMap(uuid -> {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:639]: why nest the flatmap inside the .then ?

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 5]
```diff
@@ -1127,18 +1132,18 @@ public void setActionsServerBaseUri(String actionsServerBaseUri) {
         Objects.requireNonNull(loadFlowStatus);
         Objects.requireNonNull(loadFlowParameters);
         return Mono.fromCallable(() -> {
-            StudyEntity studyEntity = new StudyEntity(null, userId, studyName, LocalDateTime.now(ZoneOffset.UTC), networkUuid, networkId, description, caseFormat, caseUuid, casePrivate, isPrivate, loadFlowStatus, loadFlowResult, loadFlowParameters, securityAnalysisUuid);
+            StudyEntity studyEntity = new StudyEntity(uuid, userId, studyName, LocalDateTime.now(ZoneOffset.UTC), networkUuid, networkId, description, caseFormat, caseUuid, casePrivate, isPrivate, loadFlowStatus, loadFlowResult, loadFlowParameters, securityAnalysisUuid);
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:1135]: is it worth it to avoid the select here ?

File: src/main/java/org/gridsuite/study/server/repository/StudyEntity.java [hunk 6]
```diff
@@ -29,7 +29,6 @@
 public class StudyEntity implements BasicStudyEntity {
 
     @Id
-    @GeneratedValue(strategy  =  GenerationType.AUTO)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/repository/StudyEntity.java:32]: is it worth it to avoid the select when creating a new study ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 7]
```diff
@@ -1115,7 +1073,7 @@ public void testCreationWithErrorBadCaseFile() throws Exception {
             Message<byte[]> message = output.receive(1000);
             assertEquals("", new String(message.getPayload()));
             MessageHeaders headers = message.getHeaders();
-            assertEquals("newStudy", headers.get(StudyService.HEADER_STUDY_NAME));
+            //assertEquals("newStudy", headers.get(StudyService.HEADER_STUDY_NAME));
```
Comment (jonenst) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:1076]: delete dead code. Or are we not ready yet to remove the studyname from the message ?

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 1 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyController.java:85]: it's used by assertUserAllowed function

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 2 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyController.java:104]: same

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 1 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyController.java:85]: getCurrentUserStudy function needs that parameter

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 2 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyController.java:104]: assertUserAllowed function needs that parameter

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 8]
```diff
@@ -102,308 +100,282 @@ public StudyController(StudyService studyService) {
...
     @ApiOperation(value = "delete the study")
     @ApiResponse(code = 200, message = "Study deleted")
-    public ResponseEntity<Mono<Void>> deleteStudy(@PathVariable("studyName") String studyName,
-                                                  @PathVariable("userId") String userId,
+    public ResponseEntity<Mono<Void>> deleteStudy(@PathVariable("studyUuid") UUID studyUuid,
                                                   @RequestHeader("userId") String headerUserId) {
-        return ResponseEntity.ok().contentType(MediaType.APPLICATION_JSON).body(studyService.assertUserAllowed(userId, headerUserId)
-                .doOnSuccess(s -> studyService.deleteStudyIfNotCreationInProgress(studyName, userId).subscribe()));
+        return ResponseEntity.ok().contentType(MediaType.APPLICATION_JSON).body(
+                studyService.deleteStudyIfNotCreationInProgress(studyUuid, headerUserId));
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyController.java:109]: Why make this synchronous ? why was it asynchronous before ?

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 9]
```diff
@@ -617,48 +632,51 @@ public void doDeleteStudyIfNotCreationInProgress(String studyName, String userId
...
     }
 
-    public Mono<StudyInfos> renameStudy(String studyName, String userId, String newStudyName) {
-        return Mono.fromCallable(() -> self.doRenameStudy(studyName, userId, newStudyName))
+    public Mono<CreatedStudyBasicInfos> renameStudy(UUID studyUuid, String userId, String newStudyName) {
+        return Mono.fromCallable(() -> self.doRenameStudy(studyUuid, userId, newStudyName))
                 .switchIfEmpty(Mono.error(new StudyException(STUDY_NOT_FOUND)))
-                .map(StudyService::toStudyInfos)
-                .doOnSuccess(s -> emitStudyChanged(studyName, StudyService.UPDATE_TYPE_STUDIES));
+                .map(StudyService::toCreatedStudyBasicInfos)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:673]: why change the API ?

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 10]
```diff
@@ -102,308 +100,282 @@ public StudyController(StudyService studyService) {
...
     }
 
-    @PostMapping(value = "/{userId}/studies/{studyName}/rename")
+    @PostMapping(value = "/studies/{studyUuid}/rename")
     @ApiOperation(value = "Update the study name", produces = "application/json")
     @ApiResponses(value = {@ApiResponse(code = 200, message = "The updated study")})
-    public ResponseEntity<Mono<StudyInfos>> renameStudy(@RequestHeader("userId") String headerUserId,
-                                                        @PathVariable("studyName") String studyName,
-                                                        @PathVariable("userId") String userId,
+    public ResponseEntity<Mono<CreatedStudyBasicInfos>> renameStudy(@RequestHeader("userId") String headerUserId,
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyController.java:355]: why change the api (same question in StudyService)

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 11]
```diff
@@ -221,6 +219,7 @@ private static CreatedStudyBasicInfos toCreatedStudyBasicInfos(StudyEntity entit
                 .studyUuid(entity.getId())
                 .caseFormat(entity.getCaseFormat())
                 .studyPrivate(entity.isPrivate())
+                .description(entity.getDescription())
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:222]: should be in another PR

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 12]
```diff
@@ -237,56 +236,63 @@ private static CreatedStudyBasicInfos toCreatedStudyBasicInfos(StudyEntity entit
...
-                                                 description, t.getT2(), caseUuid, false, LoadFlowStatus.NOT_DONE, null,  toEntity(loadFlowParameters), null);
-                          })
-                        .subscribeOn(Schedulers.boundedElastic())
-                        .doOnError(throwable -> LOGGER.error(throwable.toString(), throwable))
-                        .doFinally(r -> deleteStudyIfNotCreationInProgress(studyName, userId).subscribe())
-                        .subscribe()
-                );
+        Mono<BasicStudyInfos> basicStudyInfosMono = insertStudyCreationRequest(studyName, userId, isPrivate)
+                .map(StudyService::toBasicStudyInfos);
+
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:241]: why change all this code ?

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 13]
```diff
@@ -237,56 +236,63 @@ private static CreatedStudyBasicInfos toCreatedStudyBasicInfos(StudyEntity entit
...
-                        .subscribe()
-                );
-    }
-
-    public Mono<StudyInfos> getCurrentUserStudy(String studyName, String userId, String headerUserId) {
-        Mono<StudyEntity> studyMono = getStudyWithPreFetchedLoadFlowResult(studyName, userId);
+        Mono<BasicStudyInfos> basicStudyInfosMono = insertStudyCreationRequest(studyName, userId, isPrivate)
+                .map(StudyService::toBasicStudyInfos);
+
+        return basicStudyInfosMono.doOnSuccess(s -> importCase(caseFile, studyName)
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:259]: why change this code

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 14]
```diff
@@ -300,52 +306,61 @@ public StudyEntity doGetStudyWithPreFetchedLoadFlowResult(String studyName, Stri
     }
 
     @Transactional
-    public StudyEntity doGetStudyWithPreFetchedLoadFlowResultAndUpdateIsPrivate(String studyName, String userId, boolean toPrivate) {
-        StudyEntity studyEntity = doGetStudyWithPreFetchedLoadFlowResult(studyName, userId);
+    public StudyEntity doGetStudyWithPreFetchedLoadFlowResultAndUpdateIsPrivate(UUID studyUuid, String headerUserId, boolean toPrivate) {
+        StudyEntity studyEntity = doGetStudyWithPreFetchedLoadFlowResult(studyUuid);
         if (studyEntity != null) {
+            //only the owner of a study can change the access rights
+            if (!headerUserId.equals(studyEntity.getUserId())) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:313]: in another PR

Comment (jonenst) [reply src/main/java/org/gridsuite/study/server/StudyService.java:313]: And add a unit test

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 15]
```diff
@@ -300,52 +306,61 @@ public StudyEntity doGetStudyWithPreFetchedLoadFlowResult(String studyName, Stri
...
     @Transactional
-    public void doDeleteStudyIfNotCreationInProgress(String studyName, String userId) {
-        Optional<StudyCreationRequestEntity> studyCreationRequestEntity = studyCreationRequestRepository.findByUserIdAndStudyName(userId, studyName);
+    public void doDeleteStudyIfNotCreationInProgress(UUID uuid, String userId) {
+        Optional<StudyCreationRequestEntity> studyCreationRequestEntity = studyCreationRequestRepository.findById(uuid);
         if (studyCreationRequestEntity.isEmpty()) {
-            studyRepository.deleteByUserIdAndStudyName(userId, studyName);
+            studyRepository.findById(uuid).ifPresent(s -> {
+                if (!s.getUserId().equals(userId)) {
+                    throw new StudyException(NOT_ALLOWED);
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:339]: in another PR. And add a unit test

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 16]
```diff
@@ -617,48 +632,51 @@ public void doDeleteStudyIfNotCreationInProgress(String studyName, String userId
...
+           emitStudyChanged(studyUuid, UPDATE_TYPE_LOADFLOW)
         );
     }
 
     @Transactional
-    public StudyEntity doRenameStudy(String studyName, String userId, String newStudyName) {
-        return studyRepository.findByUserIdAndStudyName(userId, studyName).map(studyEntity -> {
+    public StudyEntity doRenameStudy(UUID studyUuid, String userId, String newStudyName) {
+        return studyRepository.findById(studyUuid).map(studyEntity -> {
+            if (!studyEntity.getUserId().equals(userId)) {
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/StudyService.java:662]: in another pr and add a unit test

File: src/main/java/org/gridsuite/study/server/dto/CreatedStudyBasicInfos.java [hunk 17]
```diff
@@ -23,4 +23,6 @@
 @ApiModel("Basic study attributes after creation succeeded ")
 public class CreatedStudyBasicInfos extends BasicStudyInfos {
     String caseFormat;
+
+    String description;
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/dto/CreatedStudyBasicInfos.java:27]: in another PR

File: src/main/java/org/gridsuite/study/server/dto/StudyInfos.java [hunk 18]
```diff
@@ -20,9 +20,6 @@
 @Getter
 @ApiModel("Study attributes")
 public class StudyInfos extends CreatedStudyBasicInfos {
-
-    String description;
-
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/dto/StudyInfos.java:25]: in another pr

Comment (jonenst) [reply src/main/java/org/gridsuite/study/server/dto/StudyInfos.java:25]: and add unit test

File: src/main/java/org/gridsuite/study/server/repository/StudyCreationRequestRepository.java [hunk 19]
```diff
@@ -24,6 +24,8 @@
 @Repository
 public interface StudyCreationRequestRepository extends JpaRepository<StudyCreationRequestEntity, UUID> {
 
+    List<StudyCreationRequestEntity> findAllByUserId(String userId);
```
Comment (jonenst) [inline src/main/java/org/gridsuite/study/server/repository/StudyCreationRequestRepository.java:27]: is it worth it to define this, instead of the findByPrimaryKey ? What's the difference

File: src/test/java/org/gridsuite/study/server/utils/MatcherStudyInfos.java [hunk 20]
```diff
@@ -38,7 +40,6 @@ protected MatcherStudyInfos(StudyInfos val) {
     @Override
     public boolean matchesSafely(StudyInfos s) {
         return super.matchesSafely(s)
-                && reference.getDescription().equals(s.getDescription())
```
Comment (jonenst) [inline src/test/java/org/gridsuite/study/server/utils/MatcherStudyInfos.java:41]: other pr

File: src/test/java/org/gridsuite/study/server/utils/MatcherCreatedStudyBasicInfos.java [hunk 21]
```diff
@@ -36,6 +40,7 @@ protected MatcherCreatedStudyBasicInfos(T val) {
     public boolean matchesSafely(T s) {
         return super.matchesSafely(s)
                 && reference.getCaseFormat().equals(s.getCaseFormat())
+                && reference.getDescription().equals(s.getDescription())
```
Comment (jonenst) [inline src/test/java/org/gridsuite/study/server/utils/MatcherCreatedStudyBasicInfos.java:43]: other pr

File: src/main/java/org/gridsuite/study/server/StudyService.java (hunk 15 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyService.java:339]: this change is related to this PR

File: src/main/java/org/gridsuite/study/server/StudyService.java (hunk 14 above)
Comment (ChamseddineBhd) [reply src/main/java/org/gridsuite/study/server/StudyService.java:313]: this change is related to this PR

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 8 above)
Comment (jonenst) [reply src/main/java/org/gridsuite/study/server/StudyController.java:109]: discussed together, we need to get the user id from the database anyway now, so in the same transaction delete the study

**Review by jonenst — APPROVED** [review]
